### PR TITLE
fix(local-runtime): stop recommending models with multi-minute prompt waits

### LINKED ISSUE #109606: [Bug]: Local Models speed gate models decode only — the prompt prefill it omits is 87-93% of an agent turn's wall clock
### Bug Description

The Local Models recommendation gates "will this be usable?" on a **decode-only** speed proxy, while the dominant cost of an agent turn is the prompt **prefill** that the proxy does not model at all.

`predicted_decode_tok_s()` is the whole model (`hermes_cli/local_runtime/catalog.py:179-186`, tree `1c671bea`):

```python
def predicted_decode_tok_s(entry, variant, budget, *, spilled=False) -> float:
    """Memory-bound decode prediction for ordering and floor-gating."""
    bandwidth = (_HOST_BANDWIDTH_GB_S if spilled
                 else _UMA_BANDWIDTH_GB_S if budget.uma
                 else _DISCRETE_BANDWIDTH_GB_S)
    bytes_per_token = max(1.0, variant.size_bytes * entry.decode_fraction)
    return bandwidth * 1e9 / bytes_per_token
```

`PLEASANT_FLOOR_TOK_S = 20.0` (`:176`) is compared against that value at `:209`, inside `recommended_entry()`. The constant's own comment states what it is for:

> Below this predicted decode speed a model stops feeling pleasant for **agentic use** (roughly reading speed with headroom for tool-call bursts).

For an agent host, decode is not the dominant term. Before a single token is decoded, the whole system prompt — system instructions, skills index, tool schemas — must be prefilled. Measured on this machine, one agent turn prefills **19,643 tokens**, which the engine processes in **224.74 s** (Vulkan). That is **87%** of the turn's wall clock. On the CPU build, which is what an AMD/Intel host resolves to by default (#100740), the same prompt extrapolates to **~595 s ≈ 10 min**, i.e. **93%** of the turn.

None of this is an input to anything. Searching the whole `local_runtime` package for a prefill throughput notion finds only memory accounting (`mtp_prefill` logits buffers in `context_policy.py:114-133`), a comment in `growth.py:69`, and a progress readout in `load_progress.py:133`. There is no prefill model. So the gate answers *"will the answer stream pleasantly?"* while the user is choosing between models on *"will the turn finish?"* — and on this machine those two answers differ by an order of magnitude.

Finally, the value the gate does compute reaches no surface. `predicted_decode_tok_s` has exactly one call site in the repository (`catalog.py:206`, inside `recommended_entry`); the only user-facing statement about speed is the Recommended badge's tooltip, which is hover-only and — by an explicit code comment — qualitative on purpose. The same sentence is produced whether the prediction is 12.8 tok/s or 3.1 tok/s.

**Related but distinct.** #102865 reports the same subsystem failing in the *opposite* direction (the auto-pick returned a config that did not fit → swap thrash → freeze); this report is about a pick that fits and is unusable for a different reason. #109557 is about the pane's *wording* (budget flags rendered as hardware claims) — this one is about the metric underneath the pick. Neither covers it.

### Steps to Reproduce

1. Use a machine with no discrete GPU (here: AMD Radeon 780M iGPU, 30.96 GB RAM) and `local_runtime.enabled: true`. The runtime resolves to the CPU llama.cpp build — see #100740 for why.
2. Open **Settings → Local Models**. Note the recommended entry and hover its badge; the tooltip is the only speed-related statement the pane makes.
3. Run one agent turn against that local model, then read the engine's own timing line:

```
$ grep -E "prompt processing" <hermes-home>/logs/llama-server.log | tail -2
[64167] slot print_timing: id 1 | task 142 | prompt processing, n_tokens = 19643, progress = 1.00, t = 224.74 s / 87.40 tokens per second
[62672] slot print_timing: id 3 | task 0   | prompt processing, n_tokens = 19643, progress = 1.00, t = 238.87 s / 82.23 tokens per second
```

The second line is the same prompt after a model reload — the cost is paid again, it is not a one-off.

4. **Or reproduce without running anything**: `catalog.py:179-186` is a pure function of `budget.uma`, `variant.size_bytes` and `entry.decode_fraction`. There is no prompt-size term and no prefill term, so no input to the recommendation can represent the cost measured above.

### Expected Behavior

The speed axis should be able to express what a turn actually costs on this machine, at least to an order of magnitude:

- Gate on a **reference-turn estimate** (time-to-first-token, and/or total turn time) rather than on decode speed alone; or
- If the floor is deliberately a *streaming-feel* proxy, then the turn cost still needs to reach the user somewhere, because that is the quantity being chosen between when they pick a local model.

Concretely on this machine: a recommendation that says "no model reaches full speed; this one comes closest" currently covers both ~4 minutes per turn (Vulkan) and ~10 minutes per turn (CPU build) with the same sentence.

### Actual Behavior

- The recommendation on this machine is `fastest-resident` — the branch that means *no candidate clears the 20 tok/s floor at all*.
- Its tooltip is the same string regardless of how far short the pick falls.
- The prefill that dominates every turn is not modelled, measured, or reported anywhere: **87%** of a turn on Vulkan, **93%** on the default CPU build.
- Measured numbers behind the above (all from this machine; the engine's own timings):

| quantity | value | how |
| --- | --- | --- |
| fixed prompt of one agent turn | **19,643 tokens** | `llama-server.log`, two independent turns |
| prefill of that prompt, Vulkan | **224.74 s / 87.40 tok/s** | `llama-server.log:1242` |
| prefill of the same prompt, second turn after reload | **238.87 s / 82.23 tok/s** | `llama-server.log:1595` |
| prefill rate, CPU build (1,768-token prompt) | 33.32 tok/s | bench |
| prefill rate, CPU — stability across sizes | 30.32 / 30.63 / 33.32 tok/s (23 / 26 / 1,768 tokens) | bench |
| prefill rate, Vulkan (same 1,768-token prompt) | 123.41 tok/s | bench |
| decode, CPU build / Vulkan | 6.2-7.0 / 8.5-9.3 tok/s | bench |
| **the value the gate evaluates** | **12.8 tok/s predicted** | `catalog.py` |
| prefill for 19,643 tokens on the CPU build | **~595 s (extrapolated)** | from the stable 33 tok/s above — extrapolation, not a measurement |

### Affected Component

Setup / Installation — managed local runtime: the model recommendation (`hermes_cli/local_runtime/catalog.py`) and the Local Models pane that surfaces it.

### Messaging Platform

N/A — desktop/CLI feature, not gateway-related.

### Debug Report

Not uploaded. `hermes debug share` publishes config plus recent logs to a third-party paste service, which for this report would include unrelated session content well beyond the bug's scope. The information the bundle would contribute is provided inline instead:

```
Hermes Agent   v0.21.2 (2026.9.11) · upstream 1c671bea, install method: git
Python         3.11.16
OS             Windows 11 x64
Hardware       AMD Radeon 780M iGPU (no discrete GPU), 30.96 GB RAM
Budget         uma=True, usable 23.07 GiB (28.83 GiB RAM x 0.80)
llama.cpp      tag b10679; cpu build by default, vulkan build installed manually
Model measured gemma4-v2-Q4_K_M.gguf, 7,365,558,464 bytes, ctx 221184
Recommended    qwen3.8-27b (UD-Q4_K_M, 15.33 GiB) - larger than the model measured
```

Happy to run `hermes debug share` and paste the links if you want the full bundle.

### Operating System

Windows 11 x64

### Python Version

3.11.16

### Hermes Version

Hermes Agent v0.21.2 (2026.9.11) · upstream `1c671bea`, install method: git

### Additional Logs / Traceback (optional)

No traceback — this is a metric that evaluates successfully and answers the wrong question. The complete implementation is three lines:

```python
# hermes_cli/local_runtime/catalog.py
169:  _DISCRETE_BANDWIDTH_GB_S = 1000.0   # representative GDDR6X/GDDR7 class
170:  _UMA_BANDWIDTH_GB_S = 210.0         # measured on unified-memory NVIDIA
171:  _HOST_BANDWIDTH_GB_S = 80.0         # spilled weights stream over host DRAM
176:  PLEASANT_FLOOR_TOK_S = 20.0
209:  pleasant = [t for t in resident if speed(t) >= PLEASANT_FLOOR_TOK_S]
```

### Root Cause Analysis (optional)

1. **The comment and the implementation disagree about what the floor measures.** The comment names the use case ("agentic use"); the expression it guards evaluates decode throughput only.
2. **The dominant term is absent from the package.** `prefill` appears in `local_runtime/` only as memory accounting (`context_policy.py:114-133`), a comment (`growth.py:69`) and a progress readout (`load_progress.py:133`). Nothing estimates how long a prompt of N tokens takes to prefill.
3. **Hermes already knows the missing input and does not use it.** The fixed overhead is assembled by Hermes itself (system prompt + skills index + tool schemas — the same quantities `hermes prompt-size` reports). The recommendation path never receives it: `recommended_entry(budget, entries)` has no prompt dimension.
4. **The existing design invites exactly this extension.** The comment above the constants says the class values are stand-ins — "A measured per-machine bandwidth could replace these class constants without touching the rule" — so the same treatment applies to a prefill rate.

Consequence, stated narrowly: the gate cannot distinguish "this machine is fine" from "every turn costs ten minutes", and the only user-facing text it produces is the same sentence in both cases.

### Proposed Fix (optional)

Smallest change that restores the floor's stated meaning — deliberately not prescribing thresholds:

**(a) Give the gate the prompt size, and a prefill axis to price it with.**
- `recommended_entry(budget, entries, prompt_tokens=...)`, defaulting to a module constant for a reference agent prompt.
- Add `_*_PREFILL_TOK_S` alongside the bandwidth constants (or a measured probe, per the comment above).
- Derive a reference turn: `ttft_s = prompt_tokens / prefill_tok_s`; `turn_s = ttft_s + answer_tokens / decode_tok_s`.
- Gate on a turn-time (or TTFT) budget instead of decode speed alone. The ordering rule's shape does not change — only which quantity it gates on.

**(b) If the floor is intentionally a streaming-feel proxy**, the turn cost still has to reach the user somewhere: the recommendation surface is the only place a local model is chosen, and it currently says nothing about it.

The reference turn size and the acceptable wall clock are maintainer calls, not something I should set from one machine's measurements.

### Are you willing to submit a PR for this?

No — reporting only.

### PR DESCRIPTION
**Pull request by fangliquanflq**
## What does this PR do?

Local Models now prices the prompt prefill and the answer stream as one representative agent turn before recommending a model. The model list also shows an estimated time to first token for each fitting catalog entry, so a user can distinguish a responsive local model from one that may wait several minutes before producing output.

### Symptom

The recommendation could select a model based only on predicted decode throughput. On machines where a roughly 20K-token Hermes prompt takes minutes to prefill, the picker and its tooltip did not represent the dominant wait before the first output token.

### Impact

Integrated-GPU and CPU-class local runtimes could receive a recommendation that looked usable even though each uncached agent turn spent most of its wall time in prompt processing.

### Bug Cause

**Trigger:** `hermes_cli/local_runtime/catalog.py` / `recommended_entry()` comparing candidates only against `predicted_decode_tok_s()`.

**Causal chain:**
1. Local Models probes a machine budget and asks the catalog for its recommended entry.
2. The catalog gates candidates on answer decode throughput without pricing the prompt that must be processed first.
3. The desktop shows a qualitative recommendation reason, so materially different multi-minute waits are indistinguishable.

**Why it is wrong:** decode throughput describes streaming feel after generation starts, not time to first token or total turn latency. For Hermes agent prompts, prefill can dominate the turn.

**Working sibling / contrast:** runtime load progress can report prompt processing after a model is already running, but the pre-download catalog decision previously had no prefill estimate to compare candidates or inform the user.

**Ruled out:** this is not a memory-fit error. `select_variant()` correctly distinguishes resident, spilled, and refused models; the missing term is latency after a model fits.

### Fix

- Estimate prefill throughput from the same hardware-class bandwidth and active-weight axes used by decode, including measured batch reuse.
- Gate quality recommendations on both the decode floor and a 20K-prompt, 512-output reference-turn budget; when no candidate clears both, rank the fallback by total reference-turn latency.
- Return estimated TTFT and total-turn seconds in each fitting catalog row and render TTFT as a visible model pill with a detailed, explicitly non-benchmark tooltip.
- Cover the prefill-gated recommendation invariant and the desktop estimate display.

## Related Issue

Closes #109606

## Why this PR vs existing PR #109618

PR #109618 adds an optional `CatalogEntry.prefill_tok_s`, but no production catalog entry supplies that field and the production catalog route does not pass `input_tokens` to `recommended_entry()`. Its `use_latency` branch therefore never runs for real Local Models rows, leaving the reported decode-only recommendation unchanged. This PR closes that production path in `catalog.py`, `web_routers/local_models.py`, and the desktop model list: every fitting catalog candidate receives a hardware-class reference-turn estimate, the recommendation actually gates on it, and users can see the resulting TTFT estimate before downloading.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)

## Changes Made

- `hermes_cli/local_runtime/catalog.py` - estimate prefill and reference-turn latency, then include it in recommendation gating and fallback ranking.
- `hermes_cli/web_routers/local_models.py` - expose reference workload, TTFT, and total-turn estimates for fitting catalog rows.
- `apps/desktop/src/app/settings/local-models-settings.tsx` - show the estimated TTFT and reference-turn details for each fitting model.
- `apps/desktop/src/types/hermes.ts`, `apps/desktop/src/i18n/` - type and describe the new estimate fields.
- `tests/hermes_cli/test_local_recommendation.py`, `apps/desktop/src/app/settings/local-models-settings.test.tsx` - cover the recommendation and presentation contracts.

## How to Test

1. Open Settings -> Local Models and inspect a fitting model row.
2. Confirm the row shows a TTFT estimate and its tooltip identifies the 20K prompt and 512-token output reference turn.
3. Run the focused Python recommendation tests and desktop component test listed below.

```bash
scripts/run_tests.sh tests/hermes_cli/test_local_recommendation.py tests/hermes_cli/test_catalog_variants.py
cd apps/desktop && npx vitest run src/app/settings/local-models-settings.test.tsx
```

### Code

- [x] I've read the Contributing Guide
- [x] My commit messages follow Conventional Commits
- [x] I searched for existing PRs and compared this implementation with PR #109618
- [x] My PR contains only changes related to this fix
- [x] I've added tests for my changes
- [x] I've considered cross-platform impact; the estimate consumes the existing platform-neutral hardware budget

### Documentation & Housekeeping

- [x] Relevant policy and API behavior are documented in code; no user guide update is required
- [x] `cli-config.yaml.example` is N/A because no config key changed
- [x] `CONTRIBUTING.md` and `AGENTS.md` are N/A because no architecture or workflow changed
- [x] Tool descriptions and schemas are N/A because no model tool changed

### CODE REVIEW

**Review by gaoanze888 — COMMENTED** [review]
The units and UX labeling are coherent, but exact head `be0a48ed895d328651ad60bad6f421ddbc6cd54d` has a deterministic decision-matrix regression: after applying the new 120-second TTFT threshold, 256/512 GB UMA recommendations change from `qwen3.8-flash-next` to `qwen3.6-35b-a3b`, while existing tests still expect Flash Next. Two matrix cases fail.

Update the expected recommendation tests in this PR and explain the intended behavior change. A later commit `6f1e791831db01188ec7e11b440c3bd5a57c0095` reportedly adjusts them, but it is not part of the reviewed head.

Residual: the 20K prompt, 120-second threshold, bandwidth values, and batch-reuse factor are empirical constants validated only for internal consistency. Consider benchmark fixtures or versioned calibration data so future tuning does not silently reshape recommendations.

**Comment by fangliquanflq** [comment]
Thanks for checking the decision matrix. The two UMA expectation failures were valid at `be0a48ed895d328651ad60bad6f421ddbc6cd54d` and are already addressed in the current head:

- `6f1e791831db01188ec7e11b440c3bd5a57c0095` changes the 256/512 GB UMA expectations to `qwen3.6-35b-a3b` with `speed-gated-quality` and documents why: Flash Next clears the decode floor but exceeds the reference-turn budget at UMA bandwidth.
- `c17038610e01218fce8b072ff9b963343e1acbbb` adds the invariant that the selected model clears the reference-turn budget while every higher-quality resident candidate exceeds it, plus route coverage for the estimate fields.
- The calibration-data suggestion is useful follow-up work, but it is not required to close this regression: the UI explicitly labels the values as a hardware-class estimate rather than a benchmark, and the complete hardware decision table makes future recommendation changes review-visible.

Verified at current head:
- `scripts/run_tests.sh tests/hermes_cli/test_local_recommendation.py tests/hermes_cli/test_local_models_routes.py` — 37 passed.
- `npx vitest run src/app/settings/local-models-settings.test.tsx` — 17 passed.

**Review by gaoanze888 — COMMENTED** [review]
Rechecked exact head `c17038610e01218fce8b072ff9b963343e1acbbb`. The intended recommendation change is now explicit and internally enforced: 256/512 GB UMA selects `qwen3.6-35b-a3b`, the selected candidate clears the reference-turn budget, and each higher-quality resident candidate is proved over budget. The API exposes the same estimator values used for selection and the UI labels them as hardware-class estimates rather than benchmark results.

Focused recommendation/route suites pass 37/37 and `git diff --check` is clean. The empirical constants remain a future calibration concern, but I found no remaining blocker for this regression.

**Comment by fangliquanflq** [comment]
Thanks for rechecking exact head `c17038610e01218fce8b072ff9b963343e1acbbb`. No additional change is needed for this review:

- The previously valid 256/512 GB UMA expectation mismatch is addressed: both cases now select `qwen3.6-35b-a3b` with `speed-gated-quality`.
- `test_recommendation_prices_prefill_before_quality` verifies that the selected model clears the reference-turn budget while every higher-quality resident candidate exceeds it.
- The catalog route exposes the estimator inputs/results, and the desktop labels the display as a hardware-class estimate rather than a benchmark.
- The calibration-data point remains useful follow-up work, but it does not identify a remaining regression in this PR.

Verification: the desktop component suite passes 17/17 at this head. The focused Python suites were previously verified 37/37 at this same head; this inbox run could not rerun them because the checkout has no pytest-capable virtualenv.
